pkg/smartcontract/nef/nef.go

```diff
@@ -0,0 +1,171 @@
+// +------------+-----------+------------------------------------------------------------+
+// | Checksum   | 4 bytes   | Sha256 of the header (CRC)                                 |
+// +------------+-----------+------------------------------------------------------------+
+// | Script     | Var bytes | Var bytes for the payload                                  |
+// +------------+-----------+------------------------------------------------------------+
+
+const (
+	// magic is a magic File header constant
+	magic uint32 = 0x3346454E
+
+	// compilerFieldSize is the length of `Compiler` File header field in bytes
+	compilerFieldSize = 32
+	// maxScriptLength s the maximum allowed contract script length
+	maxScriptLength = 1024 * 1024
+)
+
+// File represents compiled contract file structure according to the NEF3 standard.
+type File struct {
+	Header   Header
+	Checksum uint32
+	Script   []byte
+}
+
+// Header represents File header
+type Header struct {
+	Magic      uint32
+	Compiler   string
+	Version    Version
+	ScriptHash util.Uint160
+}
+
+// Version represents compiler version
+type Version struct {
+	Mayor    int32
+	Minor    int32
+	Build    int32
+	Revision int32
+}
+
+// NewFile returns new NEF3 file with script specified.
+func NewFile(script []byte) File {
+	file := File{
+		Header: Header{
+			Magic:    magic,
+			Compiler: "neo-go",
+			Version: Version{
+				Mayor:    1, // TODO
+				Minor:    1,
+				Build:    1,
+				Revision: 1,
+			},
+			ScriptHash: hash.Hash160(script),
+		},
+		Script: script,
+	}
+	file.Checksum = file.Header.CalculateChecksum()
+	return file
+}
+
+// EncodeBinary implements io.Serializable interface.
+func (v *Version) EncodeBinary(w *io.BinWriter) {
+	w.WriteU32LE(uint32(v.Mayor))
+	w.WriteU32LE(uint32(v.Minor))
+	w.WriteU32LE(uint32(v.Build))
+	w.WriteU32LE(uint32(v.Revision))
+}
+
+// DecodeBinary implements io.Serializable interface.
+func (v *Version) DecodeBinary(r *io.BinReader) {
+	v.Mayor = int32(r.ReadU32LE())
+	v.Minor = int32(r.ReadU32LE())
+	v.Build = int32(r.ReadU32LE())
+	v.Revision = int32(r.ReadU32LE())
+}
+
+// EncodeBinary implements io.Serializable interface.
+func (h *Header) EncodeBinary(w *io.BinWriter) {
+	w.WriteU32LE(h.Magic)
+	w.WriteFixedString(h.Compiler, compilerFieldSize)
+	h.Version.EncodeBinary(w)
+	h.ScriptHash.EncodeBinary(w)
+}
+
+// DecodeBinary implements io.Serializable interface.
+func (h *Header) DecodeBinary(r *io.BinReader) {
+	h.Magic = r.ReadU32LE()
+	if h.Magic != magic {
+		r.Err = errors.New("invalid magic")
+		return
+	}
+	h.Compiler = r.ReadFixedString(compilerFieldSize)
+	h.Version.DecodeBinary(r)
+	h.ScriptHash.DecodeBinary(r)
+}
+
+// CalculateChecksum returns first 4 bytes of SHA256(Header) converted to uint32.
+func (h *Header) CalculateChecksum() uint32 {
+	buf := io.NewBufBinWriter()
+	h.EncodeBinary(buf.BinWriter)
+	if buf.Err != nil {
+		panic(buf.Err)
+	}
+	return binary.LittleEndian.Uint32(hash.Sha256(buf.Bytes()).BytesBE())
```

Comment:
`hash.Checksum`?

Reply:
Or is it hashing only once?!

Reply:
Once only (see https://github.com/neo-project/neo/blob/v3.0.0-preview2/src/neo/SmartContract/NefFile.cs#L125). I also checked whether C# code is able to read properly ours `.nef` files, so it is. And all verification checks in [Deserialize](https://github.com/neo-project/neo/blob/v3.0.0-preview2/src/neo/SmartContract/NefFile.cs#L81) are passed.

Reply:
Oh, my... Yet another strange thing.